### ISSUE TITLE
Annotate `is_bearable` to return TypeGuard

### ISSUE THREAD

**Issue by justinchuby** [beartype/beartype#164]
Looks like `TypeGuard` is suitable as a return type for `is_bearable` for it to support mypy. I haven't tested. Maybe it does support mypy already?

**Comment by leycec** [comment]
**Resolved by db2b98e.** Thanks so much for the gentle reminder. I'd gradually forgotten about PEP 647 and `typing.TypeGuard`, because it's only supported only Python ≥ 3.10. But have no fear! Python can do anything that it probably shouldn't, so we forcibly made this work by transparently falling back to `bool` under Python < 3.10. Dynamically interpreted languages for the win.

Unrelatedly, it turns out our ANSI stripping routine was indeed busted. I'm unclear why exactly that blew up tonight – but it did, spectacularly. But have no fear! I forcibly made that work too by removing a single excessive backslash. If you're curious, here's the working version (...*which now resides in `beartype._util.text.utiltextcolour`, just because*):

```python
def strip_text_ansi(text: str) -> str:
    # Glory be to the one liner that you are about to read.
    return sub(r'\033[^m]*m', '', text)
```

**\o/**

**Comment by rsokl** [comment]
> so we forcibly made this work by transparently falling back to bool under Python < 3.10

Any reason why you don't just rely on `typing_extensions` in that case? I use type guards all of the time in Python < 3.10!

**Comment by leycec** [comment]
**Excellent question.** Rampant laziness is the answer, of course. I'd also kinda rather not add `typing_extensions` as a mandatory dependency – but I wonder if there isn't a way to trivially inform static type-checkers that the `typing(_extensions|).TypeGuard` type hint that we are using can transparently derive from either `typing` *or* `typing_extensions` as needed?

It's trivial for us to do that dynamically, of course – but then static type-checkers are left biting the dust. There already seems too be a bit too much dynamism in that segment of our codebase. Mypy understands the following, for example, but pyright can't seem to cope:

```python
# In "beartype.door._doorcheck":
if IS_PYTHON_AT_LEAST_3_10:
    # Defer version-specific imports.
    from beartype.typing import TypeGuard  # pyright: ignore[reportGeneralTypeIssues]

    def is_bearable(  # pyright: ignore[reportGeneralTypeIssues]
        # Mandatory flexible parameters.
        obj: object,
        hint: _T,

        # Optional keyword-only parameters.
        *, conf: BeartypeConf = BeartypeConf(),
    ) -> TypeGuard[_T]:
        ...
# Else, this interpreter targets Python < 3.10 and thus fails to supports PEP
# 647. In this case, fallback to a PEP 647-agnostic signature.
else:
    def is_bearable(  # type: ignore[misc]
        # Mandatory flexible parameters.
        obj: object,
        hint: object,

        # Optional keyword-only parameters.
        *, conf: BeartypeConf = BeartypeConf(),
    ) -> bool:
        ...
```

I'm a bit concerned already that pyright isn't correctly applying the expected type narrowing there.

![this is us](https://memegenerator.net/img/instances/65510121/she-canna-take-any-more-captain-shes-gonna-blow.jpg)

**Comment by rsokl** [comment]
Type checkers are special cased to always know what `typing_extensions` is even if it isn't installed (it's treated like part of the standard library)[^1]. The following should work:

```python
if typing.TYPE_CHECKING:
    from typing_extensions import TypeGuard
else:
    try:
        from typing import TypeGuard
    except ImportError:
        try:
            from typing_extensions import TypeGuard
        except ImportError:
            TypeGuard = bool
```

[^1]: [I learned this the hard way from a mypy core dev](https://github.com/HypothesisWorks/hypothesis/issues/3397#issuecomment-1218539697)

**Comment by rsokl** [comment]
> I'm a bit concerned already that pyright isn't correctly applying the expected type narrowing there.

Actually, <sub><sup>your precious</sup></sub> mypy isn't doing any type-narrowing either. It is just greedily taking the first path (also it raises an error, see below). See that in the following, depending on your Python env, either `f` or `g` should type-check as `def (x: builtins.str) -> Any`, but mypy doesn't show this:
```python
import sys

IS_PYTHON_AT_LEAST_3_11 = sys.version_info >= (3, 11)
IS_PYTHON_AT_LEAST_3_10 = IS_PYTHON_AT_LEAST_3_11 or sys.version_info >= (3, 10)


if IS_PYTHON_AT_LEAST_3_10:
    def f(x: int): ...
else:
    def f(x: str): ...

if not IS_PYTHON_AT_LEAST_3_10:  # note the *not*
    def g(x: int): ...
else:
    def g(x: str): ...

reveal_type(f)  # mypy: Revealed type is "def (x: builtins.int) -> Any"
reveal_type(g)  # mypy: Revealed type is "def (x: builtins.int) -> Any"
```
```
main.py:10: error: All conditional function variants must have identical signatures
main.py:15: error: All conditional function variants must have identical signatures
main.py:17: note: Revealed type is "def (x: builtins.int) -> Any"
main.py:18: note: Revealed type is "def (x: builtins.int) -> Any"
Found 2 errors in 1 file (checked 1 source file)
```

If you modify the code to use explicit version-info checks, you get the appropriate behavior from mypy:

```python
import sys
if sys.version_info >= (3, 10):
    def f(x: int): ...
else:
    def f(x: str): ...

if not sys.version_info >= (3, 10):  # note the *not*
    def g(x: int): ...
else:
    def g(x: str): ...

reveal_type(f)  # mypy: Revealed type is "def (x: builtins.int) -> Any"
reveal_type(g)  # mypy: Revealed type is "def (x: builtins.str) -> Any"
```
```
main.py:12: note: Revealed type is "def (x: builtins.str) -> Any"
main.py:13: note: Revealed type is "def (x: builtins.int) -> Any"
Success: no issues found in 1 source file
```

**Comment by leycec** [comment]
**Holy balls of AI glory.** Welp, that's why MIT pays you the big beaver bucks while I just scrounge in filth in a cabin in the woods routinely raided by beavers.

I *totally* grok you on the `IS_PYTHON_AT_LEAST_3_10` versus `sys.version_info >= (3, 10)` issue. Tragically, tuple comparisons incur a mostly negligible cost that becomes non-negligible when you do them often enough. @beartype internally gates so much logic behind interpreter version checks that we do them often enough. Mypy, I blame you for your own crass stupidity! :sweat: 

That said...

```python
if typing.TYPE_CHECKING:
    from typing_extensions import TypeGuard
else:
    try:
        from typing import TypeGuard
    except ImportError:
        try:
            from typing_extensions import TypeGuard
        except ImportError:
            TypeGuard = bool
```

**This is unabashed brilliance.** I'd *totally* forgotten about the classic `typing.TYPE_CHECK` nuclear option in my rush for the `beartype 0.11.0` glory. I'll immediately refactor our current approach from crude `IS_PYTHON_AT_LEAST_3_10` checks to your wise and eldritch alternative. Thanks so much, @rsokl! Take all my home-distilled maple syrup, for today you have earned nature's gold.

**Comment by rsokl** [comment]
My pleasure! Always happy to help where I can 😄

**Comment by leycec** [comment]
@rsokl: Your brilliance is a beatific light that guides @beartype into hitherto uncharted dark waters of type narrowing. Commit 85ef6684 applies the [*extremely* clever circumvention you outlined above](https://github.com/beartype/beartype/issues/164#issuecomment-1242895830). Because mypy and `pyright` are both now quiete about `beartype.door.is_bearable()`, I desperately suspect that we're now narrowing your types properly.

Thanks so much for all your AI-like cleverness. Let us know if there's something explicit you need from @beartype for [`hydra-zen`](https://github.com/mit-ll-responsible-ai/hydra-zen), [Hypothesis](https://hypothesis.readthedocs.io/en/latest/), or anything else in your vast stable of exciting projects. Until then, our immediate plan is to push `beartype 0.11.0` out with these changes and more... *this Friday.*

I say that every Friday. Let's see if I actually keep a promise this time. :sweat_smile:

**Comment by justinchuby** [comment]
Just wondering (without having tested myself) does `bool[T]` work without errors when TypeGuard is bool?

**Comment by leycec** [comment]
*...heh.* You got me. Pretty sure that's busted as currently defined under Python < 3.10 when `typing_extensions` is *not* installed. Since resolving that's trivial on our end, this is on my immediately Gods-help-us `FIXMEORBUST:` list for tonight.

Kinda was hoping nobody would notice that in the interim. Of course, everybody did. :face_exhaling:

**Comment by leycec** [comment]
In theory, this has finally been **resolved at long last by commit 2ed18dcc.** In practice, our arduous and painful journey of unspeakable woe is probably only beginning.

If someone who is not me wouldn't mind testing this, I'd be eternally indebted. Until then, I'll happily pretend this works as advertised. Blessed sleep, take me now! :sleeping:
